### PR TITLE
Add fast-path construction to `DAGCircuit` methods

### PR DESCRIPTION
**Pull request by jakelishman**
### Summary

This optimises the constructor functions
`DAGCircuit.apply_operation_back`, `apply_operation_front` and `compose` by giving callers a way to skip the validity checking when passing known-good values.  In most cases when we're operating on the DAG, we are certain that we're already upholding the invariants of the data structure by nature of having a `copy_empty_like` or similar, and the user should not pay a runtime price for what we should catch during testing.

### Details and comments

I ran a benchmark suite and `asv` didn't pop up anything statistically significant in changes but quite a bit of stuff saw ~5% improvements. I'm just running benchmarks on a laptop in other use, and our benchmark suite is quite outdated now.

In a more targeted microbenchmark with setup
```python
from qiskit.circuit.library import QuantumVolume
from qiskit.converters import circuit_to_dag
qc = QuantumVolume(100, seed=0).decompose()
```
the call `circuit-to_dag(qc, copy_operations=False)` went from 34.3(7)ms to 28.6(7)ms, so a ~15% speedup.  That's indicative of what's happening when we're doing things like rebuilding the DAG after Sabre or during `ApplyLayout`.

### CODE REVIEW

File: qiskit/dagcircuit/dagcircuit.py [hunk 1]
```diff
@@ -592,23 +608,6 @@ def _decrement_op(self, op):
...
-        Returns:
-            int: The integer node index for the new op node on the DAG
-        """
-        # Add a new operation node to the graph
-        new_node = DAGOpNode(op=op, qargs=qargs, cargs=cargs, dag=self)
-        node_index = self._multi_graph.add_node(new_node)
-        new_node._node_id = node_index
-        self._increment_op(op)
-        return node_index
-
```
Comment (jakelishman) [inline qiskit/dagcircuit/dagcircuit.py:611]: This is a fairly simple function used in only two places which already necessarily had a lot of duplication, and throws away information that the outer contexts need (the `DAGOpNode` - we previously had to retrieve it from `_multi_graph` again), so I just inlined it into its call sites.

**Comment by jlapeyre** [comment]
It would be useful to design a canonical way to choose whether to do expensive safety checks. The simplest "design" is using the same name for the option to elide checks. For example, we have

https://github.com/Qiskit/qiskit/blob/61e5bde38b393540c8640b320127fffad918e9f6/qiskit/synthesis/discrete_basis/solovay_kitaev.py#L93

https://github.com/Qiskit/qiskit/blob/61e5bde38b393540c8640b320127fffad918e9f6/qiskit/circuit/library/generalized_gates/linear_function.py#L78-L79

I'm concerned about adding more variety to these options, making the barrier to cleaning it up in the future even larger.

Maybe this PR could use the design we want (again, maybe simplest is best and the design is just a naming convention). And a followup PR could make this uniform to the extent possible across qiskit.

**Comment by jakelishman** [comment]
Fwiw we also have `validate`, `require_cp`, `require_tp` in `quantum_info`, and `cycle_check` within `DAGCircuit`, which are all types of input validation as well.  I don't mind especially what we go for (I err on the shorter side when the meaning is unambiguous) - which do you prefer so I can update the PR?

**Comment by jlapeyre** [comment]
Short story. I am ok with using `check` for now. It may not be the best name, but the marginal tech debt of a name that must be changed is small. Trying to figure this out now would derail this PR. 

longer:

For the sake of brevity, I didn't mention concerns about: corner cases, internal vs external api, conceptual and motivational differences in eliding checks, interactions between checks.

Of course that all becomes immediately important.  I won't suggest unifying `require_tp` and `require_cp` into `check`, as they were already split (https://github.com/Qiskit/qiskit/pull/3660) from `require_cptp` and given different defaults. No reason was given, but I'm sure there was one. I grepped all of the qiskit repos and only find these options in a few tests in `qiskit-experiments`, and always set to `False`.

I think standardizing on something like `check_some_invariant`, and `check_all` might be a good first step. I.e. in this case `check` becomes `check_all` and `require_tp` becomes `check_tp`.  Some of these currently log a warning others error. That should probably be worked into the naming.

**Comment by jakelishman** [comment]
That sounds like an interesting topic to pursue potentially under the guise of defining a "naming convention" guide for Qiskit.  Perhaps we could move that into a separate issue to track?

*edit*: just pushed a commit to fix the merge conflict.

**Comment by jlapeyre** [comment]
In `bits_in_operation`, shouldn't this
https://github.com/Qiskit/qiskit/blob/7252db3723c286f6c4b4865c654061911ff33e3b/qiskit/dagcircuit/dagcircuit.py#L572-L574
have `elif isinstance(operation, SwitchCaseOp)`? without `elif` it's harder to read if you don't know the code. I mean it's not obvious that these are not mutually exclusive branches.

File: qiskit/dagcircuit/dagcircuit.py [hunk 2]
```diff
@@ -580,6 +580,22 @@ def _bits_in_operation(operation):
...
+            operation (qiskit.circuit.Operation): the operation to check.
+        """
+        # This is separate to `_bits_in_operation` because most of the time there won't be any bits,
+        # so we want a fast path to be able to skip creating and testing a generator for emptiness.
+        #
+        # If updating this, also update `DAGCirucit._bits_in_operation`.
+        return getattr(operation, "condition", None) is not None or isinstance(
+            operation, SwitchCaseOp
+        )
+
```
Comment (jlapeyre) [inline qiskit/dagcircuit/dagcircuit.py:598]: This optimization is in addition to the optimization discussed in the opening comment, right? Did you benchmark this optimization?  It's clear that `set(self._bits_in_operation(op)).union(cargs)` will be slower than just `cargs` even when the generator is empty. But I would not be surprised if this is not measurable in benchmarks.

Comment (jlapeyre) [inline qiskit/dagcircuit/dagcircuit.py:598]: This optimization is in addition to the optimization discussed in the opening comment, right? Did you benchmark this optimization?  It's clear that `set(self._bits_in_operation(op)).union(cargs)` will be slower than just `cargs` even when the generator is empty. But I would not be surprised if this is not measurable in benchmarks.

**Comment by jlapeyre** [comment]
I think that this
https://github.com/Qiskit/qiskit/blob/7252db3723c286f6c4b4865c654061911ff33e3b/qiskit/dagcircuit/dagcircuit.py#L1194-L1197

should use `_operation_may_have_bits` as well.

**Comment by jlapeyre** [comment]
The questions above are all I have. The main part of the PR, i.e. introducing a `check` option looks good to me.

**Comment by jakelishman** [comment]
I think I wrote that `SwitchCaseOp` `if`/`if` as a defensive thing - the two branches are mutually exclusive right now, but to me making it `if`/`else` felt like making more assumptions than were needed for the function.

File: qiskit/dagcircuit/dagcircuit.py (hunk 2 above)
Comment (jakelishman) [reply qiskit/dagcircuit/dagcircuit.py:598]: Yeah, looks like I didn't mention this little bit in the commit message.  Removing the `_may_have_bits` check and always eagerly calling `_bits_in_operation` caused the microbenchmark at the top to go from 27.7(7)ms with the current PR to 29.3(9)ms when timed just now.

**Comment by jakelishman** [comment]
I've done the `substitute_node_with_dag` thing in 14b215cd4.  In general I didn't look to optimise any of the other `DAGCircuit` functions to keep the scope of the PR targetted and because DAG reconstruction from `SabreSwap` is one place that bottlenecks `transpile` which only uses `apply_operation_back` at the moment.

File: qiskit/dagcircuit/dagcircuit.py (hunk 2 above)
Comment (jlapeyre) [reply qiskit/dagcircuit/dagcircuit.py:598]: That's worth the trouble although it adds a bit of clutter

**Comment by jakelishman** [comment]
(you need to leave an approving review to enable merge - the merge queue handles bringing the PR up-to-date with `main`, so that's not necessary, but the code-owner approval is)

**Comment by jlapeyre** [comment]
Huh. I tried to do that. Clicked on the green approve button....
I'll try again.
Maybe the first click only changed input focus ?

**Review by jlapeyre — APPROVED** [review]
Looks good.

**Review by jlapeyre — APPROVED** [review]
Looks good.

**Pull request merged**